semgrep-core/src/fixing/tests/Unit_autofix_printer.ml

```diff
@@ -0,0 +1,105 @@
+ * or fix pattern, just that we lift the correct number of characters. The tests
+ * in `semgrep-core/tests` exercise the system as a whole, and ensure that the
+ * fixes tested there are correct.
+ *
+ * TODO generalize the test runner once we have multiple languages to test
+ *)
+let test_python_autofix_printer () =
+  let check (expected, target, pattern, fix_pattern) =
+    Common2.with_tmp_file ~str:target ~ext:"py" (fun target_file ->
+        (* Make a fake string to use when printing instead of the target. Fill
+         * it with 't' so that we can identify it in the final output. *)
+        let fake_target_contents = String.make (String.length target) 't' in
+        (* Same as above, but for the fix pattern *)
+        let fake_fix_pattern = String.make (String.length fix_pattern) 'p' in
+        let matches =
+          Unit_engine.match_patterns ~lang:Lang.Python
+            ~hook:(fun _ -> ())
+            ~file:target_file ~pattern ~fix_pattern:(Some fix_pattern)
+        in
+        (* To keep it simple, we make sure that each example here has only a
+         * single match. *)
+        let match_ =
+          match matches with
+          | [ m ] -> m
+          | lst ->
+              failwith
+                (spf
+                   "wrong number of matches for `%s`. expected exactly 1, got \
+                    %d"
+                   target (List.length lst))
+        in
+        let fix_pattern_ast =
+          Parse_pattern.parse_pattern Lang.Python fix_pattern
+        in
+        let metavars = match_.env in
+        let fixed_pattern_ast =
+          match
+            Autofix_metavar_replacement.replace_metavars metavars
+              fix_pattern_ast
+          with
+          | Ok x -> x
+          | Error e ->
+              failwith (spf "Failed to replace metevars for `%s`:\n%s" target e)
+        in
+        let fix_text =
+          match
+            Autofix_printer.print_ast ~lang:Lang.Python ~metavars
+              ~target_contents:(lazy fake_target_contents)
+              ~fix_pattern_ast ~fix_pattern:fake_fix_pattern fixed_pattern_ast
+          with
+          | Ok x -> x
+          | Error e ->
+              failwith (spf "Failed to print for test case `%s`:\n%s" target e)
+        in
+        (* Replace the fake target contents with the rendered fix *)
+        let start, end_ =
+          let start, end_ = match_.Pattern_match.range_loc in
+          let _, _, end_charpos = Parse_info.get_token_end_info end_ in
+          (start.Parse_info.charpos, end_charpos)
+        in
+        let full_fixed_text =
+          let before = Str.first_chars fake_target_contents start in
+          let after = Str.string_after fake_target_contents end_ in
+          before ^ fix_text ^ after
+        in
+        Alcotest.(check string) fix_pattern expected full_fixed_text)
+  in
+  List.iter check
+    [
+      (* TODO foobar should be lifted from the pattern *)
+      ("foobar", "foo", "foo", "foobar");
```

Comment:
maybe we could define a record for those tests, so it's easier to remember what each tuple element correspond to.
Maybe short names like {expexted: "foobar"; tgt: "foo"; pat:"foo"; fix"foobar"}